### PR TITLE
Convert auth service to use the new Auth0 SPA

### PR DESCRIPTION
**Pull request by nigel-wells**
- Converted `AuthService` to use the new Auth0 SPA package
- Enable use of refresh tokens which then automatically uses the offline scope
- Fixed tests setup to better replicate different logging in flows
- Improved handling of return URL to avoid redirect loops
- Used a new dedicated callback URL to know when to check with auth0 on handling the callback code

Updates made to auth0 configuration
- Added `http://localhost:5000/callback` as a new callback URL
- Removed old beta callback URLs
- Refresh token rotation enabled
- API access settings | enabled Allow Offline Access

There appears to be a few different circumstances to trigger an update to tokens:
- API token settings expiry is hit - this appear to be our main use of the `expires_at` setting
- Application ID Token expiration

Both of those are stored and handled but the auth0 package and it will automatically renew them as required. We currently have a renewal timer set on the API expiration so that we can force a refresh early enough before we send the access token to the backend. The default for the ID Token expiration is 36000 or 10 hours. I currently have both of these set to 300 or 5 minutes. 30 seconds prior to this a silent token exchange will take place and the expiration will be reset to another 5 minutes. All going well there will be no redirect take place as it was at this point browsers that blocked 3rd party cookies would return a `login_required` response.

The expected behavior after switching to the new auth flow is:
- If the browser is already running the app then nothing will change. On an expired token if the browser supports 3rd party cookies then it will continue to authenticate fine. If the browser does not support 3rd party cookies it will get the usual `login_required` response sending you back to auth0. Auth0 will likely assume you're already logged in and send you back to SF. SF will not have the new token data required from the new auth0 package and will attempt to log the user in again but this time with the correct callback and scope. Auth0 will now prompt for the `offline_access` scope and then redirect back to SF where you left off.
- If the browser is reloaded, and you're already logged in, then it will be redirected to auth0 for the new `offline_access` scope

---
This change is [Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1375)

### CODE REVIEW

**Review by RaymondLuong3 — CHANGES_REQUESTED** [review]
Good work! And the test coverage is good too. I have some suggested improvements, but the critical parts appear to be there.

> Reviewed 4 of 10 files at r1, 7 of 7 files at r2.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-:-N2xB90F1KUDMBFR3mzq:b-9d5u3o)* status: 10 of 11 files reviewed, 5 unresolved discussions (waiting on @nigel-wells and @RaymondLuong3)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts` line 64 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-N2wIggoE6sddpfAGgaI:-N2wIggp8BKKf1atcSzZ:b55msu0) ([raw file](https://github.com/sillsdev/web-xforge/blob/0625f2654f8542793a4ec3e9e3bb2301c1a5df63/src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts#L64)):*
> ```XML
>   private renewTokenPromise?: Promise<void>;
>   private checkSessionPromise?: Promise<GetTokenSilentlyVerboseResponse | null>;
>   private readonly auth0 = this.auth0Service.init({
> ```

Can you add a type for this?



_Code quote:_
```XML
 private readonly auth0 = this.auth0Service.init({
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts` line 160 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-N2wET9QErqTDu6dds4O:-N2wET9QErqTDu6dds4P:bc8kats) ([raw file](https://github.com/sillsdev/web-xforge/blob/0625f2654f8542793a4ec3e9e3bb2301c1a5df63/src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts#L160)):*
> ```XML
>     try {
>       return await this.auth0!.getTokenSilently();
>     } catch {
> ```

We typically want to avoid using the bang to assert not null. We could return undefined if `this.auth0` is undefined. Something like `return await this.auth0?.getTokenSilently()` 
This and below.



_Code quote:_
```XML
      return await this.auth0!.getTokenSilently();
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts` line 329 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-N2wLq6L-ERpXiKymQgw:-N2wLq6L-ERpXiKymQgx:b-2lzl1i) ([raw file](https://github.com/sillsdev/web-xforge/blob/0625f2654f8542793a4ec3e9e3bb2301c1a5df63/src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts#L329)):*
> ```XML
>           )
>           .then(async () => {
>             const withoutState = clone(authDetails);
> ```

It would be helpful if you leave a comment here explaining why you are removing the state.



_Code quote:_
```XML
       const withoutState = clone(authDetails);
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts` line 348 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-N2wM59rD8ADYDTyT3QD:-N2wM59rD8ADYDTyT3QE:b-8s707x) ([raw file](https://github.com/sillsdev/web-xforge/blob/0625f2654f8542793a4ec3e9e3bb2301c1a5df63/src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts#L348)):*
> ```XML
>     // Ensure the return URL is one we want to return the user to
>     if (this.isValidReturnUrl(state.returnUrl)) {
>       this.router.navigateByUrl(state.returnUrl!, { replaceUrl: true });
> ```

Instead of asserting not null with the bang, you can add a null check in the iff above.



_Code quote:_
```XML
      this.router.navigateByUrl(state.returnUrl!, { replaceUrl: true });
    } else {
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts` line 356 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-N2wN-mEClmPF66bdoXw:-N2wN-mFBuBBhSPPT4zc:b98zots) ([raw file](https://github.com/sillsdev/web-xforge/blob/0625f2654f8542793a4ec3e9e3bb2301c1a5df63/src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts#L356)):*
> ```XML
> 
>   private isValidReturnUrl(returnUrl: string | undefined): boolean {
>     return !(returnUrl == null || this.isCallbackUrl(returnUrl) || returnUrl === '/login');
> ```

Why is it necessary to check that the the return url is not the callback url? Why is that invalid?



_Code quote:_
```XML
|| this.isCallbackUrl(returnUrl) || 
```

**Review by nigel-wells — COMMENTED** [review]
The coverage report above is interesting - when I view it locally `auth.service.ts` is at 90%

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-:-N2y1jb248a4HImm2RLt:b-5l4dxc)* status: 10 of 11 files reviewed, 5 unresolved discussions (waiting on @RaymondLuong3)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts` line 64 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-N2wIggoE6sddpfAGgaI:-N2y0AF33OeWqRZbq1nw:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/0625f2654f8542793a4ec3e9e3bb2301c1a5df63/src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts#L64)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Can you add a type for this?
</blockquote></details>

Done.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts` line 160 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-N2wET9QErqTDu6dds4O:-N2y0R1_7soZHL7O_7oW:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/0625f2654f8542793a4ec3e9e3bb2301c1a5df63/src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts#L160)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

We typically want to avoid using the bang to assert not null. We could return undefined if `this.auth0` is undefined. Something like `return await this.auth0?.getTokenSilently()` 
This and below.
</blockquote></details>

Done.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts` line 329 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-N2wLq6L-ERpXiKymQgw:-N2y1Ta98HEy_oicO-r8:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/0625f2654f8542793a4ec3e9e3bb2301c1a5df63/src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts#L329)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

It would be helpful if you leave a comment here explaining why you are removing the state.
</blockquote></details>

Done.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts` line 348 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-N2wM59rD8ADYDTyT3QD:-N2y0bDT-ZX5h_84yizN:be8i0ml) ([raw file](https://github.com/sillsdev/web-xforge/blob/0625f2654f8542793a4ec3e9e3bb2301c1a5df63/src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts#L348)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Instead of asserting not null with the bang, you can add a null check in the iff above.
</blockquote></details>

This one is probably ok as it has already been verified that it isn't null by `isValidReturnUrl`
___
*[`src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts` line 356 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-N2wN-mEClmPF66bdoXw:-N2y0lMnAFo6OR7wi9pJ:b5jlbdq) ([raw file](https://github.com/sillsdev/web-xforge/blob/0625f2654f8542793a4ec3e9e3bb2301c1a5df63/src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts#L356)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Why is it necessary to check that the the return url is not the callback url? Why is that invalid?
</blockquote></details>

I believe there was a bug in the current implementation where the return URL sometimes ends up as the same URL returned from auth0 which then gets the user stuck in a loop. This makes sure that doesn't happen and we can also expand upon it if we find other situations that might cause it. It may be that we need to reverse the logic and only have a set of valid paths we allow to be return urls.

**Review by nigel-wells — COMMENTED** [review]
Perhaps the one noted above is from the original commit though before I added some extra tests in though?

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-:-N2y2kpM6ahVxqV5cx9n:bizcuzm)* status: 10 of 11 files reviewed, 5 unresolved discussions (waiting on @RaymondLuong3)

**Comment by Nateowami** [comment]
@nigel-wells While working on other things with my work based off this branch, I've run into the following after coming back from Auth0 (GitHub shows it blurry; click it to see it clearly):



This happened after I used the "log in with Paratext" button in SF that is shown when SF thinks the user does not have access to Paratext. After I come back from Auth0 I get this screen. The tokens apparently don't update, because I'm still greeted with the "log in with Paratext" buttons. Those buttons trigger `linkParatext(returnUrl: string)` in `AuthService`.

**Review by RaymondLuong3 — CHANGES_REQUESTED** [review]
I tried a few more scenarios in chrome incognito mode. As for what Nathaniel pointed out, when I could proceed to my Paratext-linked account when I used paratext credentials linked to an existing SF account. The page reloaded after the pop-up, but was smooth otherwise.

> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-:-N3617nt4Va6lvCfreTo:b-9zp699)* status: all files reviewed, 1 unresolved discussion (waiting on @nigel-wells)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts` line 348 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-N2wM59rD8ADYDTyT3QD:-N35y8jODXzlA4pHiThU:b8sec31) ([raw file](https://github.com/sillsdev/web-xforge/blob/0625f2654f8542793a4ec3e9e3bb2301c1a5df63/src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts#L348)):*
<details><summary><i>Previously, nigel-wells (Nigel Wells) wrote…</i></summary><blockquote>

This one is probably ok as it has already been verified that it isn't null by `isValidReturnUrl`
</blockquote></details>

That is fine. There are some places we use the not null assertion in production code, but I want to keep that to a minimal.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts` line 356 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-N2wN-mEClmPF66bdoXw:-N35yRgvEJMLfKYF8wxE:b-igjdvb) ([raw file](https://github.com/sillsdev/web-xforge/blob/0625f2654f8542793a4ec3e9e3bb2301c1a5df63/src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts#L356)):*
<details><summary><i>Previously, nigel-wells (Nigel Wells) wrote…</i></summary><blockquote>

I believe there was a bug in the current implementation where the return URL sometimes ends up as the same URL returned from auth0 which then gets the user stuck in a loop. This makes sure that doesn't happen and we can also expand upon it if we find other situations that might cause it. It may be that we need to reverse the logic and only have a set of valid paths we allow to be return urls.
</blockquote></details>

Ok, that is god to keep in mind. I would add a comment so that it is easy to see that.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts` line 195 at r5](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-N35zVWu763sdaCTCxfH:-N35zVWu763sdaCTCxfI:b4sofsq) ([raw file](https://github.com/sillsdev/web-xforge/blob/f35050698819d3b6abbd9d4b20af5ac56e3fa1b4/src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts#L195)):*
> ```XML
>     }
>     this.unscheduleRenewal();
>     await this.auth0!.loginWithRedirect(authOptions);
> ```

I'm wondering if we even need the bang at all for `this.auth0`. Since we declare it to be a defined type, it shouldn't be necessary. Here and below.



_Code quote:_
```XML
    await this.auth0!.loginWithRedirect(authOptions);
```

**Review by Nateowami — COMMENTED** [review]
> As for what Nathaniel pointed out, when I could proceed to my Paratext-linked account when I used paratext credentials linked to an existing SF account. The page reloaded after the pop-up, but was smooth otherwise.

Nigel fixed the issue I ran into. We talked about it in a meeting but didn't comment on the PR itself.

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-:-N36K4Kl9huuNeNzQ1LT:b-wneujx)* status: all files reviewed, 1 unresolved discussion (waiting on @nigel-wells)

**Review by nigel-wells — COMMENTED** [review]
@raymondluong3 the page reload is expected behavior - at least that is what it has always done anyway if a user ID changes which is what can happen when linking PT accounts. We could probably make that smoother if we felt it occurred enough but I'd do that as a separate task.

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-:-N3Hg3321BFpDj75n30k:blk2hqj)* status: 10 of 11 files reviewed, 1 unresolved discussion (waiting on @RaymondLuong3)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts` line 348 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-N2wM59rD8ADYDTyT3QD:-N3HdGGv17vnJXzesx6f:b-q693a5) ([raw file](https://github.com/sillsdev/web-xforge/blob/0625f2654f8542793a4ec3e9e3bb2301c1a5df63/src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts#L348)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

That is fine. There are some places we use the not null assertion in production code, but I want to keep that to a minimal.
</blockquote></details>

As above, I've now removed.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts` line 356 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-N2wN-mEClmPF66bdoXw:-N3Hdnky8ukZ0BoTPhEg:bh25w3m) ([raw file](https://github.com/sillsdev/web-xforge/blob/0625f2654f8542793a4ec3e9e3bb2301c1a5df63/src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts#L356)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Ok, that is god to keep in mind. I would add a comment so that it is easy to see that.
</blockquote></details>

Have added a comment to the method
___
*[`src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts` line 195 at r5](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-N35zVWu763sdaCTCxfH:-N3Hd-7qDDBg-qwZpv8o:b8yb9q) ([raw file](https://github.com/sillsdev/web-xforge/blob/f35050698819d3b6abbd9d4b20af5ac56e3fa1b4/src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts#L195)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

I'm wondering if we even need the bang at all for `this.auth0`. Since we declare it to be a defined type, it shouldn't be necessary. Here and below.
</blockquote></details>

You're absolutely right. When I first started refactoring in the new auth0 package I had it setup as possibly undefined. Towards the end I was able to get it setup defined in the constructor. I've gone ahead and removed all the bangs.

**Review by RaymondLuong3 — APPROVED** [review]
Thanks for your work. From my testing this is working as expected. I approve.

> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-:-N3Kj3Y1EeIvdGCKKmKL:b9ihhw8)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @nigel-wells)

**Review by Nateowami — CHANGES_REQUESTED** [review]
Let's make the needed changes to the Auth0 tenants before merging this. @nigel-wells Can you list the changes that need to be made to the qa and live tenants?

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-:-N3PzxdI97PTH_aX_Bqb:bxjwrvt)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @nigel-wells)

**Review by nigel-wells — COMMENTED** [review]
At a guess, as I don't have access to the repo, would be:
## Resource server
```
{
   ...
   "allow_offline_access": true,
   ...
}
```

## Client
 ```
{
     "callbacks": [
         "http://localhost/callback/auth0" // Just needs the path updated from "projects" to "callback/auth0"
     ],
     "refresh_token": {
      ...
        "rotation_type": "rotating",
      ...
}
```

Happy to do a screenshare on this if that is easier for us both to look at how it is setup for QA/Prod.

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-:-N3QqTIw4YLaeaTo6yyo:b-pvdyjb)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @nigel-wells)

**Review by nigel-wells — COMMENTED** [review]
I think there is an issue with the linking to PT which I've discovered when testing the passwordless login - hold off merging until I investigate more

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-:-N3S3_J71YcvcZBXtObI:b-g6jzp7)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @nigel-wells)

**Review by nigel-wells — COMMENTED** [review]
I've just pushed through the latest updates for the PT linking fix. I'll start looking at the auth0 CLI config stuff now

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-:-N4-9_2J6bqFMT3sUC86:b-mkfeow)* status: 9 of 15 files reviewed, all discussions resolved (waiting on @RaymondLuong3)

**Review by RaymondLuong3 — DISMISSED** [review]
> Reviewed 6 of 6 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1375)* status: all files reviewed, 1 unresolved discussion (waiting on @nigel-wells)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts` line 333 at r7](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-N43AKQbE1NUWnOxJDXK:-N43AKQcBu2TQB4JA4rV:bm99gb6) ([raw file](https://github.com/sillsdev/web-xforge/blob/9783587ae6f0fe787a28f28f08d14cb078f4acca/src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts#L333)):*
> ```XML
>       try {
>         await this.commandService.onlineInvoke(USERS_URL, 'linkParatextAccount', { primaryId, secondaryId });
>         // Trigger the login phase again so that local auth0 switches back its understanding to the primary account
> ```

What is going on here? Can this be removed? since we are not triggering the login phase and simply checking the session?



_Code quote:_
```XML
switches back its understanding to the primary 
```

**Review by nigel-wells — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1375)* status: 14 of 15 files reviewed, 1 unresolved discussion (waiting on @RaymondLuong3)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts` line 333 at r7](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-N43AKQbE1NUWnOxJDXK:-N441OG2-dwz2-k40lfk:b-dm6cy) ([raw file](https://github.com/sillsdev/web-xforge/blob/9783587ae6f0fe787a28f28f08d14cb078f4acca/src/SIL.XForge.Scripture/ClientApp/src/xforge-common/auth.service.ts#L333)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

What is going on here? Can this be removed? since we are not triggering the login phase and simply checking the session?
</blockquote></details>

Whoops - and that is why we have code review :)

I've updated the comment as well to help it better reflect what is happening as well.

**Review by nigel-wells — COMMENTED** [review]
Just noting here that I have a feature branch ready to go for the auth0 config when it is time to start testing on QA

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-:-N44lrh98lFKKuxcjth4:bq01wrf)* status: 14 of 15 files reviewed, 1 unresolved discussion (waiting on @RaymondLuong3)

**Review by Nateowami — APPROVED** [review]
Thanks @nigel-wells. Let's :shipit: to qa.

> Reviewed 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1375#-:-N45HHex59j2tMLjDEYD:byoh1h4)* status: all files reviewed, 1 unresolved discussion (waiting on @RaymondLuong3)

**Pull request merged**
